src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c

```diff
@@ -0,0 +1,1249 @@
+        return;
+    }
+
+    if (ptr->mainloop)
+    {
+        pa_threaded_mainloop_stop(ptr->mainloop);
+    }
+
+    if (ptr->context)
+    {
+        pa_context_disconnect(ptr->context);
+        pa_context_unref(ptr->context);
+    }
+
+    if (ptr->mainloop)
+    {
+        pa_threaded_mainloop_free(ptr->mainloop);
+    }
+
+    PaUtil_FreeMemory(ptr);
+}
+
+
+static void PulseAudioCheckContextStateCb(
+    pa_context * c,
+    void *userdata
+)
+{
+    PaPulseAudioHostApiRepresentation *ptr = userdata;
+    assert(c);
+
+    /* If this is null we have big problems and we probably are out of memory */
+    if (!c)
+    {
+        PA_PULSEAUDIO_SET_LAST_HOST_ERROR(0,
+                                          "PulseAudioCheckContextStateCb: Out of memory");
+        pa_threaded_mainloop_signal(ptr->mainloop, 0);
+        return;
+    }
+
+    ptr->state = pa_context_get_state(c);
+    pa_threaded_mainloop_signal(ptr->mainloop, 0);
+}
+
+int _PulseAudioAddAudioDevice(
+    PaPulseAudioHostApiRepresentation *hostapi,
+    const char *interfaceName,
+    const char *realName,
+    int inputChannels,
+    int outputChannels,
+    double defaultLowInputLatency,
+    double defaultHighInputLatency,
+    double defaultLowOutputLatency,
+    double defaultHighOutputLatency,
+    const long defaultSampleRate
+)
+{
+    /* These should be at leat 1 */
+    int l_iRealNameLen = strnlen(realName, 1024) + 1;
+    int l_iDeviceNameLen = strnlen(interfaceName, 1024) + 1;
+    char *l_ptrName = NULL;
+    
+    hostapi->pulseaudioDeviceNames[hostapi->deviceCount] = (char *) PaUtil_GroupAllocateMemory(hostapi->allocations,
+                                                           l_iRealNameLen);
+    hostapi->deviceInfoArray[hostapi->deviceCount].name =  (char *) PaUtil_GroupAllocateMemory(hostapi->allocations,
+                                                           l_iDeviceNameLen);
+  
+    if( !hostapi->pulseaudioDeviceNames[hostapi->deviceCount] &&
+        !hostapi->deviceInfoArray[hostapi->deviceCount].name)
+    {
+       PA_PULSEAUDIO_SET_LAST_HOST_ERROR(0,
+                                         "Can't alloc memory"); 
+       return paInsufficientMemory;
+    }
+    /*PA_UNLESS(hostapi->pulseaudioDeviceNames[hostapi->deviceCount] =
+              (char *) PaUtil_GroupAllocateMemory(hostapi->allocations,
+                                                  l_iRealNameLen),
+              paInsufficientMemory);*/
+
+    /* PA_UNLESS(hostapi->deviceInfoArray[hostapi->deviceCount].name =
+              (char *) PaUtil_GroupAllocateMemory(hostapi->allocations,
+                                                  l_iDeviceNameLen),
+              paInsufficientMemory); */
+
+    strncpy((char *)
+            hostapi->pulseaudioDeviceNames[hostapi->deviceCount],
+            realName, l_iRealNameLen);
+
+    strncpy((char *) hostapi->deviceInfoArray[hostapi->deviceCount].name,
```

Comment:
name is "const char *name" such a remove const cast is ugly. You should save the allocated memory in a local. 

char *name = (char *) PaUtil_GroupAllocateMemory ...
strncpy(name, interfaceName ...
hostapi->deviceInfoArray[hostapi->deviceCount].name = name;